src/deepsparse/transformers/utils/decoder_kv_cache.py

```diff
@@ -0,0 +1,180 @@
+            The cache tensor is a numpy array of shape
+            [batch_size, num_heads, sequence_length - num_input_ids, hidden_size]
+        :param sequence_length: The total capacity of the kv cache.
+            This is the maximum number of tokens that can be stored
+            in the kv cache until we start to overwrite the oldest
+            entries in the cache.
+        :param freeze_first_position: If set to True, once the kv cache
+            gets filled, the position along the sequence length axis
+            that corresponds to the first token will be frozen.
+            This assures that, once the KV cache is full (there are no
+            "blank" entries), and we are removing the "oldest" entry
+            from the cache, we will nevertheless keep the cache entry
+            that corresponds to the BOS token in the sequence.
+            By default, is set to False.
+        """
+        self.session_id = session_id
+        self._state = state
+        self._freeze_first_position = freeze_first_position
+        self._total_cache_capacity = sequence_length
+
+        if self._kv_cache is not None:
+            self._kv_cache.reset(
+                num_tokens=self._total_cache_capacity,
+                freeze_first_position=[int(self._freeze_first_position)],
+            )
+
+    def update_session(
+        self,
+        state: Dict[str, Any],
+        num_tokens: int,
+        input_ids_len: int,
+    ):
+        """
+        Updating the session is identical with taking the kv cache
+        output of from the forward pass and restructuring it, so it
+        can be directly used as input for the next forward pass.
+
+        :param state: The state of the cache. This is a dictionary
+            that maps the name of the cache array to the cache array.
+            The cache tensor is a numpy array of shape
+            [batch_size, num_heads, sequence_length, hidden_size]
+        :param num_tokens: The number of tokens processed so far,
+            corresponding to the number of "non-blank" entries in the
+            kv cache array. Even though all numpy arrays in the state
+            of the cache have constant sequence length, this length
+            corresponds to the total "capacity" of the
+            cache. A portion of this memory can be "blank".
+        :param input_ids_len: The number of input ids in the current
+            input batch: (batch_size, length).
+            Corresponds to `input_ids.shape[1]`
+        """
+
+        # total_capacity = num_tokens (num of non-blank tokens) +
+        # + num_padded_entries (num of blank tokens)
+        num_padded_entries = max(0, self._total_cache_capacity - num_tokens)
+        # we want to remove input_ids_len entries from the cache
+        # because len_input_ids + inp_cache_len = out_cache_len
+        num_entries_to_delete = input_ids_len
+
+        while num_entries_to_delete > 0:
```

Comment:
calling delete each step was good for our pseudocode, but in production, might be faster to select the idxs to delete/keep here and then do a single slice later so we're not constantly writing new arrays